GRPC client fails after a call has been done with a RestClient if prometheus metrics are configured

### Describe the bug

After upgrading to Quarkus 3.34 (and also last 3.35.2), our application raises this unexpected error when we try to call a service through a GrpcClient. The issue appears only after a first call with a Rest Client and if prometheus metrics are configured.

The log messages : 

```
2026-05-13 10:13:55,867 ERROR [org.jboss.resteasy.reactive.server.handlers.PublisherResponseHandler] (vert.x-eventloop-thread-3) Exception in SSE server handling, impossible to send it to client: io.grpc.StatusRuntimeException: UNKNOWN
	at io.grpc.Status.asRuntimeException(Status.java:532)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:581)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at io.quarkus.grpc.runtime.supports.IOThreadClientInterceptor$1$1.lambda$onClose$3(IOThreadClientInterceptor.java:70)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:252)
	at io.vertx.core.impl.ContextInternal.lambda$runOnContext$0(ContextInternal.java:50)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalArgumentException: Prometheus requires that all meters with the same name have the same set of tag keys. There is already an existing meter named 'http_client_requests_seconds' containing tag keys [clientName, method, outcome, status, uri]. The meter you are attempting to register has keys [address, clientName, method, outcome, status, uri].
	at io.micrometer.prometheus.PrometheusMeterRegistry.lambda$throwExceptionOnRegistrationFailure$19(PrometheusMeterRegistry.java:619)
	at io.micrometer.core.instrument.MeterRegistry.meterRegistrationFailed(MeterRegistry.java:1291)
	at io.micrometer.prometheus.PrometheusMeterRegistry.lambda$applyToCollector$18(PrometheusMeterRegistry.java:593)
	at java.base/java.util.concurrent.ConcurrentHashMap.compute(ConcurrentHashMap.java:1940)
	at io.micrometer.prometheus.PrometheusMeterRegistry.applyToCollector(PrometheusMeterRegistry.java:579)
	at io.micrometer.prometheus.PrometheusMeterRegistry.newTimer(PrometheusMeterRegistry.java:328)
	at io.micrometer.core.instrument.MeterRegistry.lambda$timer$6(MeterRegistry.java:379)
	at io.micrometer.core.instrument.MeterRegistry.getOrCreateMeter(MeterRegistry.java:725)
	at io.micrometer.core.instrument.MeterRegistry.registerMeterIfNecessary(MeterRegistry.java:652)
	at io.micrometer.core.instrument.MeterRegistry.timer(MeterRegistry.java:377)
	at io.micrometer.core.instrument.Timer$Builder.register(Timer.java:471)
	at io.micrometer.core.instrument.Timer$Builder.register(Timer.java:465)
	at io.micrometer.core.instrument.composite.CompositeTimer.registerNewMeter(CompositeTimer.java:205)
	at io.micrometer.core.instrument.composite.CompositeTimer.registerNewMeter(CompositeTimer.java:35)
	at io.micrometer.core.instrument.composite.AbstractCompositeMeter.add(AbstractCompositeMeter.java:67)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at java.base/java.util.Collections$SetFromMap.forEach(Collections.java:5944)
	at io.micrometer.core.instrument.composite.CompositeMeterRegistry.lambda$new$0(CompositeMeterRegistry.java:67)
	at io.micrometer.core.instrument.composite.CompositeMeterRegistry.lock(CompositeMeterRegistry.java:189)
	at io.micrometer.core.instrument.composite.CompositeMeterRegistry.lambda$new$1(CompositeMeterRegistry.java:67)
	at io.micrometer.core.instrument.MeterRegistry.getOrCreateMeter(MeterRegistry.java:735)
	at io.micrometer.core.instrument.MeterRegistry.registerMeterIfNecessary(MeterRegistry.java:652)
	at io.micrometer.core.instrument.MeterRegistry.timer(MeterRegistry.java:377)
	at io.micrometer.core.instrument.Timer$Builder.register(Timer.java:471)
	at io.micrometer.core.instrument.Timer$Builder.lambda$withRegistry$0(Timer.java:453)
	at io.quarkus.micrometer.runtime.binder.vertx.VertxHttpClientMetrics$3.responseEnd(VertxHttpClientMetrics.java:195)
	at io.quarkus.micrometer.runtime.binder.vertx.VertxHttpClientMetrics$3.responseEnd(VertxHttpClientMetrics.java:108)
	at io.vertx.core.http.impl.Http2ClientConnection.metricsEnd(Http2ClientConnection.java:212)
	at io.vertx.core.http.impl.Http2ClientConnection.access$200(Http2ClientConnection.java:39)
	at io.vertx.core.http.impl.Http2ClientConnection$Stream.onEnd(Http2ClientConnection.java:322)
	at io.vertx.core.http.impl.Http2ClientConnection.onHeadersRead(Http2ClientConnection.java:206)
	at io.vertx.core.http.impl.Http2ConnectionBase.onHeadersRead(Http2ConnectionBase.java:209)
	at io.netty.handler.codec.http2.Http2FrameListenerDecorator.onHeadersRead(Http2FrameListenerDecorator.java:46)
	at io.netty.handler.codec.http2.Http2EmptyDataFrameListener.onHeadersRead(Http2EmptyDataFrameListener.java:63)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:430)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:345)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader$2.processFragment(DefaultHttp2FrameReader.java:489)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readHeadersFrame(DefaultHttp2FrameReader.java:497)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:257)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:174)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:186)
	at io.netty.handler.codec.http2.DecoratingHttp2ConnectionDecoder.decodeFrame(DecoratingHttp2ConnectionDecoder.java:61)
	at io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:402)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:462)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:545)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:484)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.vertx.core.http.impl.VertxHttp2ConnectionHandler.channelRead(VertxHttp2ConnectionHandler.java:419)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:171)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:732)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:658)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	... 4 more
```

### Expected behavior

Same behaviour than with Quarkus 3.33: the grpc call does not throw an exception

### Actual behavior

The Grpc call throws the exception (see logs below)

### How to Reproduce?

Unzip the reproducer and run `mvn test`

Change the quarkus version to test 3.33 or 3.34, etc.

[grpc-reproducer.zip](https://github.com/user-attachments/files/27698768/grpc-reproducer.zip)

### Additional information

I think the bug is related to this one: https://github.com/quarkusio/quarkus/issues/53914

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/ResteasyClientMetricsFilter.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/ResteasyClientMetricsFilter.java
@@ -79,6 +79,7 @@ public class ResteasyClientMetricsFilter implements ClientRequestFilter, ClientR
 
                 sample.stop(timer
                         .withTags(Tags.of(
+                                address(requestContext),
                                 HttpCommonTags.method(requestContext.getMethod()),
                                 HttpCommonTags.uri(requestPath, requestContext.getUri().getPath(), statusCode,
                                         httpMetricsConfig.isClientSuppress4xxErrors()),
@@ -91,6 +92,15 @@ public class ResteasyClientMetricsFilter implements ClientRequestFilter, ClientR
 
     private RequestMetricInfo getRequestMetric(ClientRequestContext requestContext) {
         return (RequestMetricInfo) requestContext.getProperty(REQUEST_METRIC_PROPERTY);
+    }
+
+    private Tag address(ClientRequestContext requestContext) {
+        String host = requestContext.getUri().getHost();
+        int port = requestContext.getUri().getPort();
+        if (host == null) {
+            return Tag.of("address", "none");
+        }
+        return Tag.of("address", port > 0 ? host + ":" + port : host);
     }
 
     private Tag clientName(ClientRequestContext requestContext) {

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/vertx/RestClientMetricsFilter.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/vertx/RestClientMetricsFilter.java
@@ -78,6 +78,7 @@ public class RestClientMetricsFilter implements ResteasyReactiveClientRequestFil
 
                 sample.stop(timer
                         .withTags(Tags.of(
+                                address(requestContext),
                                 HttpCommonTags.method(requestContext.getMethod()),
                                 HttpCommonTags.uri(requestPath, requestContext.getUri().getPath(), statusCode,
                                         httpMetricsConfig.isClientSuppress4xxErrors()),
@@ -90,6 +91,15 @@ public class RestClientMetricsFilter implements ResteasyReactiveClientRequestFil
 
     private RequestMetricInfo getRequestMetric(ClientRequestContext requestContext) {
         return (RequestMetricInfo) requestContext.getProperty(REQUEST_METRIC_PROPERTY);
+    }
+
+    private Tag address(ResteasyReactiveClientRequestContext requestContext) {
+        String host = requestContext.getUri().getHost();
+        int port = requestContext.getUri().getPort();
+        if (host == null) {
+            return Tag.of("address", "none");
+        }
+        return Tag.of("address", port > 0 ? host + ":" + port : host);
     }
 
     private Tag clientName(ResteasyReactiveClientRequestContext requestContext) {

--- a/integration-tests/grpc-interceptors/pom.xml
+++ b/integration-tests/grpc-interceptors/pom.xml
@@ -24,6 +24,10 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-rest-client</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-junit</artifactId>
             <scope>test</scope>
         </dependency>
@@ -71,6 +75,19 @@
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-rest-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-rest-client-deployment</artifactId>
             <version>${project.version}</version>
             <type>pom</type>
             <scope>test</scope>

--- a/integration-tests/grpc-interceptors/src/main/java/io/quarkus/grpc/examples/interceptors/HelloWorldEndpoint.java
+++ b/integration-tests/grpc-interceptors/src/main/java/io/quarkus/grpc/examples/interceptors/HelloWorldEndpoint.java
@@ -8,6 +8,8 @@ import jakarta.ws.rs.Path;
 import jakarta.ws.rs.PathParam;
 import jakarta.ws.rs.core.Response;
 
+import org.eclipse.microprofile.rest.client.inject.RestClient;
+
 import examples.GreeterGrpc;
 import examples.HelloReply;
 import examples.HelloRequest;
@@ -18,6 +20,10 @@ import io.smallrye.mutiny.Uni;
 @Path("/hello")
 public class HelloWorldEndpoint {
     static Set<String> invoked = new HashSet<>();
+
+    @RestClient
+    PingRestClient pingRestClient;
+
     @GrpcClient("hello")
     GreeterGrpc.GreeterBlockingStub blockingHelloService;
 
@@ -42,5 +48,19 @@ public class HelloWorldEndpoint {
     public Uni<String> helloMutiny(@PathParam("name") String name) {
         return mutinyHelloService.sayHello(HelloRequest.newBuilder().setName(name).build())
                 .onItem().transform(HelloReply::getMessage);
+    }
+
+    @GET
+    @Path("/ping")
+    public String ping() {
+        return "pong";
+    }
+
+    @GET
+    @Path("/rest-then-grpc/{name}")
+    public String restThenGrpc(@PathParam("name") String name) {
+        String pingResult = pingRestClient.ping();
+        HelloReply helloReply = blockingHelloService.sayHello(HelloRequest.newBuilder().setName(name).build());
+        return pingResult + " " + helloReply.getMessage();
     }
 }

--- a/integration-tests/grpc-interceptors/src/main/java/io/quarkus/grpc/examples/interceptors/PingRestClient.java
+++ b/integration-tests/grpc-interceptors/src/main/java/io/quarkus/grpc/examples/interceptors/PingRestClient.java
@@ -1,0 +1,15 @@
+package io.quarkus.grpc.examples.interceptors;
+
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
+
+@RegisterRestClient(configKey = "ping")
+@Path("/hello")
+public interface PingRestClient {
+
+    @GET
+    @Path("/ping")
+    String ping();
+}

--- a/integration-tests/grpc-interceptors/src/main/resources/application.properties
+++ b/integration-tests/grpc-interceptors/src/main/resources/application.properties
@@ -15,3 +15,5 @@ quarkus.grpc.clients.hello.port=9001
 
 %o2n.quarkus.grpc.clients.hello.port=8081
 %o2n.quarkus.grpc.clients.hello.use-quarkus-grpc-client=false
+
+quarkus.rest-client.ping.url=http://localhost:${quarkus.http.port:8081}

--- a/integration-tests/grpc-interceptors/src/test/java/io/quarkus/grpc/example/interceptors/HelloWorldEndpointTestBase.java
+++ b/integration-tests/grpc-interceptors/src/test/java/io/quarkus/grpc/example/interceptors/HelloWorldEndpointTestBase.java
@@ -40,6 +40,17 @@ class HelloWorldEndpointTestBase {
         assertThat(response).isEqualTo("Hello neo-mutiny");
     }
 
+    @Test
+    public void testRestClientThenGrpcDoesNotConflictWithMetrics() {
+        String response = get("/hello/rest-then-grpc/neo").then().statusCode(200)
+                .extract().asString();
+        assertThat(response).isEqualTo("pong Hello neo");
+
+        String metrics = get("/q/metrics").then().statusCode(200)
+                .extract().asString();
+        assertThat(metrics).contains("http_client_requests_seconds");
+    }
+
     public void ensureThatMetricsAreProduced() {
         String metrics = get("/q/metrics")
                 .then().statusCode(200)

--- a/integration-tests/rest-client/src/test/java/io/quarkus/it/rest/client/MultipartResourceTest.java
+++ b/integration-tests/rest-client/src/test/java/io/quarkus/it/rest/client/MultipartResourceTest.java
@@ -30,6 +30,6 @@ public class MultipartResourceTest {
                 .statusCode(200)
                 // /echo is ignored in application.properties
                 .body(not(containsString(
-                        "http_client_requests_seconds_count{clientName=\"localhost\",method=\"POST\",outcome=\"SUCCESS\",status=\"200\",uri=\"/echo\",} 1.0")));
+                        "clientName=\"localhost\",method=\"POST\",outcome=\"SUCCESS\",status=\"200\",uri=\"/echo\",} 1.0")));
     }
 }

--- a/integration-tests/rest-client/src/test/java/io/quarkus/it/rest/client/wronghost/BaseExternalWrongHostTestCase.java
+++ b/integration-tests/rest-client/src/test/java/io/quarkus/it/rest/client/wronghost/BaseExternalWrongHostTestCase.java
@@ -22,6 +22,6 @@ public abstract class BaseExternalWrongHostTestCase {
                 .then()
                 .statusCode(200)
                 .body(containsString(
-                        "http_client_requests_seconds_count{clientName=\"localhost\",method=\"GET\",outcome=\"SUCCESS\",status=\"200\",uri=\"root\"}"));
+                        "clientName=\"localhost\",method=\"GET\",outcome=\"SUCCESS\",status=\"200\",uri=\"root\"}"));
     }
 }

